language/bytecode_verifier/src/instantiation_loops.rs

```diff
@@ -0,0 +1,202 @@
+
+        fn rec(ty: &SignatureToken) -> Vec<TypeParameterIndex> {
+            match ty {
+                Bool | Address | U64 | String | ByteArray => vec![],
+                TypeParameter(idx) => vec![*idx],
+                Reference(ty) | MutableReference(ty) => rec(ty),
+                Struct(_, tys) => tys.iter().map(rec).flatten().collect(),
+            }
+        }
+
+        let mut type_params = rec(ty);
+        type_params.sort();
+        type_params.dedup();
+        type_params
+    }
+
+    /// Builds a graph G such that
+    ///   - Each type formal of a generic function is a node in G.
+    ///   - There is an edge from type formal f_T to g_T if f_T is used to instantiate g_T in a
+    ///     call.
+    ///     - Each edge is labeled either `Identity` or `TyConApp`. See `Edge` for details.
+    fn build_graph(&mut self) {
+        for def in self
+            .module
+            .function_defs()
+            .iter()
+            .filter(|def| !def.is_native())
+        {
+            for instr in &def.code.code {
+                if let Bytecode::Call(other_func_idx, type_actuals_idx) = instr {
+                    let other_func_handle = self.module.function_handle_at(*other_func_idx);
+                    if other_func_handle.module != ModuleHandleIndex::new(0) {
+                        continue;
+                    }
+
+                    let type_actuals = &self.module.locals_signature_at(*type_actuals_idx).0;
+                    for (formal_idx, ty) in type_actuals.iter().enumerate() {
+                        let formal_idx = formal_idx as TypeParameterIndex;
+                        let node_to = self.get_or_add_node(Node(*other_func_idx, formal_idx));
+                        if let SignatureToken::TypeParameter(actual_idx) = ty {
+                            let node_from = self.get_or_add_node(Node(def.function, *actual_idx));
+                            self.graph.add_edge(node_from, node_to, Edge::Identity);
+                        } else {
+                            let type_params = Self::extract_type_parameters(ty);
+                            for type_param in type_params {
+                                let node_from =
+                                    self.get_or_add_node(Node(def.function, type_param));
+                                self.graph.add_edge(node_from, node_to, Edge::TyConApp(&ty));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /// Computes the strongly connected components of the graph built and keep the ones that
+    /// contain at least one `TyConApp` edge. Such components indicate there exists a loop such
+    /// that an input type can get "bigger" infinitely many times along the path, also creating
+    /// infinitely many types. This is precisely the kind of constructs we want to forbid.
+    fn find_non_trivial_loops(&self) -> Vec<(Vec<NodeIndex>, Vec<EdgeIndex>)> {
```

Comment:
How is what this function returns a loop?  Please characterize what this function is returning more precisely.